File: package/umt_python/tests/unit/math/test_clamp.py

import unittest
from src.math import clamp

class TestClamp(unittest.TestCase):
    def test_should_return_the_value_when_within_range(self):
        self.assertEqual(clamp(5, 0, 10), 5)

    def test_should_return_the_min_when_value_is_below_range(self):
        self.assertEqual(clamp(-3, 0, 10), 0)

    def test_should_return_the_max_when_value_is_above_range(self):
        self.assertEqual(clamp(15, 0, 10), 10)

    def test_should_return_min_when_value_equals_min(self):
        self.assertEqual(clamp(0, 0, 10), 0)

    def test_should_return_max_when_value_equals_max(self):
        self.assertEqual(clamp(10, 0, 10), 10)

    def test_should_handle_negative_ranges(self):
        self.assertEqual(clamp(-5, -10, -1), -5)
        self.assertEqual(clamp(0, -10, -1), -1)
        self.assertEqual(clamp(-20, -10, -1), -10)

    def test_should_handle_floating_point_values(self):
        self.assertEqual(clamp(0.5, 0, 1), 0.5)
        self.assertEqual(clamp(1.5, 0, 1), 1)
        self.assertEqual(clamp(-0.5, 0, 1), 0)

    def test_should_handle_min_equal_to_max(self):
        self.assertEqual(clamp(5, 3, 3), 3)
        self.assertEqual(clamp(1, 3, 3), 3)

File: package/umt_rust/tests/math/test_clamp.rs
use umt_rust::math::umt_clamp;

#[test]
fn test_should_return_the_value_when_within_range() {
    assert_eq!(umt_clamp(5.0, 0.0, 10.0), 5.0);
}

#[test]
fn test_should_return_the_min_when_value_is_below_range() {
    assert_eq!(umt_clamp(-3.0, 0.0, 10.0), 0.0);
}

#[test]
fn test_should_return_the_max_when_value_is_above_range() {
    assert_eq!(umt_clamp(15.0, 0.0, 10.0), 10.0);
}

#[test]
fn test_should_return_min_when_value_equals_min() {
    assert_eq!(umt_clamp(0.0, 0.0, 10.0), 0.0);
}

#[test]
fn test_should_return_max_when_value_equals_max() {
    assert_eq!(umt_clamp(10.0, 0.0, 10.0), 10.0);
}

#[test]
fn test_should_handle_negative_ranges() {
    assert_eq!(umt_clamp(-5.0, -10.0, -1.0), -5.0);
    assert_eq!(umt_clamp(0.0, -10.0, -1.0), -1.0);
    assert_eq!(umt_clamp(-20.0, -10.0, -1.0), -10.0);
}

#[test]
fn test_should_handle_floating_point_values() {
    assert_eq!(umt_clamp(0.5, 0.0, 1.0), 0.5);
    assert_eq!(umt_clamp(1.5, 0.0, 1.0), 1.0);
    assert_eq!(umt_clamp(-0.5, 0.0, 1.0), 0.0);
}

#[test]
fn test_should_handle_min_equal_to_max() {
    assert_eq!(umt_clamp(5.0, 3.0, 3.0), 3.0);
    assert_eq!(umt_clamp(1.0, 3.0, 3.0), 3.0);
}
